src/main/java/org/llorllale/cactoos/matchers/Assertion.java

```diff
@@ -71,38 +71,59 @@
  * @todo #18:30min Replace all uses of MatcherAssert.assertThat() with
  *  Assertion, and ban all overloads of the former in forbidden-apis.txt.
  *  We should also look into banning common matchers like Matchers.is(), etc.
- * @todo #18:30min Assertion relies on the operation under test to be idempotent
- *  in order to match "error matchers" such as `Throws` as expected. This may
- *  not be the case for all operations and could be a problem. Investigate if
- *  there's a way around/through this.
  */
 @SuppressWarnings("PMD.AvoidCatchingGenericException")
 public final class Assertion<T> {
+
     /**
-     * Reason for refuting this assertion.
+     * The results of the testing scenario.
      */
-    private final String reason;
+    private final Scalar<Boolean> failed;
+
     /**
-     * Behaviour to test.
+     * The description of the testing scenario.
      */
-    private final Scalar<T> test;
+    private final Scalar<Description> scenario;
+
     /**
-     * Matcher against which the behaviour will be tested.
+     * Ctor.
+     * @param reason Reason for refuting this assertion
+     * @param test The behaviour to test
+     * @param matcher Matcher to test behaviour
      */
-    private final Matcher<T> matcher;
+    public Assertion(
+        final String reason, final Scalar<T> test, final Matcher<T> matcher
+    ) {
+        this(
+            () -> !matcher.matches(test.value()),
+            new Scenario(reason, test, matcher)
+        );
+    }
 
     /**
      * Ctor.
      * @param reason Reason for refuting this assertion
      * @param test The behaviour to test
      * @param matcher Matcher to test behaviour
      */
+    @SuppressWarnings("unchecked")
     public Assertion(
-        final String reason, final Scalar<T> test, final Matcher<T> matcher
+        final String reason, final Scalar<T> test, final Throws<T> matcher
     ) {
-        this.reason = reason;
-        this.test = test;
-        this.matcher = matcher;
+        this(
+            () -> !matcher.matches(test),
+            new Scenario(reason, test, (Matcher<T>) matcher)
+        );
+    }
+
+    /**
+     * Ctor.
+     * @param failed The failed status of testing
+     * @param scenario The description of testing scenario
+     */
+    private Assertion(final Scalar<Boolean> failed, final Scenario scenario) {
```

Comment:
@dgroup the second parameter should be `Scalar<Boolean>`. Just use the highest abstraction possible.

Reply:
@llorllale, you mean `Scalar<Description>`, right?